include/devicetree.h

```diff
@@ -1,15 +1,1283 @@
+ * the individual cells in each interrupt specifier are named "irq" and
+ * "priority" by the node's binding, here are some examples:
+ *
+ *     #define SERIAL DT_NODELABEL(my_serial)
+ *
+ *     Example usage                    Value
+ *     -------------                    -----
+ *     DT_IRQ_IDX(SERIAL, 0, irq)          33
+ *     DT_IRQ_IDX(SERIAL, 0, priority)      0
+ *     DT_IRQ_IDX(SERIAL, 1, irq,          34
+ *     DT_IRQ_IDX(SERIAL, 1, priority)      1
+ *
+ * @param node_id node identifier
+ * @param idx logical index into the interrupt specifier array
+ * @param cell cell name specifier
+ * @return the named value at the specifier given by the index
+ */
+#define DT_IRQ_IDX(node_id, idx, cell)   \
+	DT_CAT(node_id, _IRQ_IDX_##idx##_VAL_##cell)
+
+/**
+ * @brief Get a value within an interrupt specifier by name
+ *
+ * It might help to read the argument order as being similar to
+ * "node->interrupts.name.cell".
+ *
+ * This can be used to get information about an individual interrupt
+ * when a device generates more than one, if the bindings give each
+ * interrupt specifier a name.
+ *
+ * @param node_id node identifier
+ * @param name lowercase-and-underscores interrupt specifier name
+ * @param cell cell name specifier
+ * @return the named value at the specifier given by the index
+ */
+#define DT_IRQ_NAME(node_id, name, cell) \
+	DT_CAT(node_id, _IRQ_NAME_##name##_VAL_##cell)
+
+/**
+ * @brief Get an interrupt specifier's value
+ * Equivalent to DT_IRQ_IDX(node_id, 0, cell).
+ * @param node_id node identifier
+ * @param cell cell name specifier
+ * @return the named value at that index
+ */
+#define DT_IRQ(node_id, cell) DT_IRQ_IDX(node_id, 0, cell)
+
+/**
+ * @brief Get a node's (only) irq number
+ *
+ * Equivalent to DT_IRQ(node_id, irq). This is provided as a convenience
+ * for the common case where a node generates exactly one interrupt,
+ * and the IRQ number is in a cell named "irq".
+ *
+ * @param node_id node identifier
+ * @return the interrupt number for the node's only interrupt
+ */
+#define DT_IRQN(node_id) DT_IRQ(node_id, irq)
+
+/**
+ * @}
+ */
+
+/**
+ * @defgroup devicetree-generic-chosen Chosen nodes
+ * @{
+ */
+
+/**
+ * @brief Get a node identifier for a /chosen node property
+ * This is only valid to call if DT_HAS_CHOSEN_NODE(prop) is 1.
+ * @param prop lowercase-and-underscores property name for
+ *             the /chosen node
+ * @return a node identifier for the chosen node property
+ */
+#define DT_CHOSEN(prop) DT_CAT(DT_CHOSEN_, prop)
+
+/**
+ * @}
+ */
+
+/**
+ * @defgroup devicetree-generic-exist Existence checks
+ * @{
+ */
+
+/**
+ * @brief Does a node identifier refer to a usable node?
+ *
+ * Example uses:
+ *
+ *     DT_HAS_NODE(DT_PATH(soc, i2c@12340000))
+ *     DT_HAS_NODE(DT_ALIAS(an_alias_name))
+ *
+ * Tests whether a node identifier refers to a node which:
+ *
+ * - exists,
+ * - is enabled (has status property "okay"), and
+ * - has a matching binding
```

Comment:
Is the test conjunctive or disjunctive of these conditions?  Presumably conjunctive, since it can't be enabled if it doesn't exist, and IIRC is an error to have a node that doesn't correspond to a binding (or is that changed now, so we can add arbitrary stuff without validation?).  In any case this seems a little confusing, especially since the implementation seems to only reference "exists".

Another point for the future: per #19448 it's likely that very soon we're going to need to be able to get information about nodes where status is not "okay" but one of the values that indicates it can be enabled at runtime.

Reply:
> Is the test conjunctive or disjunctive of these conditions? Presumably conjunctive

Conjunctive is correct: "exists, is enabled [...], and". Will try to clarify. Pay no attention to the macro name behind the curtain ;).

> Another point for the future: per #19448 it's likely that very soon we're going to need to be able to get information about nodes where status is not "okay" but one of the values that indicates it can be enabled at runtime.

Seems like this can be supported at that time via a new `DT_NODE_HAS_STATUS(node_id, SOME_SPECIAL_TOKEN)`.